script.embuary.helper/resources/lib/helper.py

```diff
@@ -227,29 +252,224 @@ def json_call(method,properties=None,sort=None,query_filter=None,limit=None,para
+
+            tags[label] = {'type': 'movies', 'id': str(tagid)}
+            all_tags.append(label)
+            duplicate_handler.append(label)
+
+    except KeyError:
+        pass
+
+    try:
+        for tag in tvshow_tags['result']['tags']:
+            label, tagid = tag['label'], tag['tagid']
+
+            if label in tag_blacklist:
+                continue
+
+            if label not in duplicate_handler:
+                tags[label] = {'type': 'tvshows', 'id': str(tagid)}
+                all_tags.append(label)
+            else:
+                tags[label] = {'type': 'mixed', 'id': str(tagid)}
+
+    except KeyError:
+        pass
+
+    all_tags.sort()
+    winprop('library.tags.all', get_joined_items(all_tags))
+
+    return tags
+
+
+def set_library_tags(tags,whitelist=None,save=True,clear=False):
+    setting = 'tags_whitelist.' + xbmc.getSkinDir() +'.data'
+    index = 0
+
+    if tags and not clear:
+        if not whitelist:
+            try:
+                whitelist = eval(addon_data('tags_whitelist.' + xbmc.getSkinDir() +'.data'))
+            except Exception:
+                pass
+
+        for item in tags:
+            if item in whitelist:
+                winprop('library.tags.%d.title' % index, item)
+                winprop('library.tags.%d.type' % index, tags[item].get('type'))
+                winprop('library.tags.%d.id' % index, tags[item].get('id'))
+                index += 1
+
+    for clean in range(index,30):
+        winprop('library.tags.%d.title' % clean, clear=True)
+        winprop('library.tags.%d.type' % clean, clear=True)
+        winprop('library.tags.%d.id' % clean, clear=True)
+
+    whitelist.sort()
+    winprop('library.tags', get_joined_items(whitelist))
+
+    if save:
+        addon_data('tags_whitelist.' + xbmc.getSkinDir() +'.data', str(whitelist))
+
+
+def addon_data_cleanup(number_of_days=60):
+    time_in_secs = time.time() - (number_of_days * 24 * 60 * 60)
+
+    ''' Image storage maintaining. Deletes all created images which were unused in the
+        last 60 days. The image functions are touching existing files to update the
+        modification date. Often used images are never get deleted by this task.
+    '''
+    for file in os.listdir(ADDON_DATA_IMG_PATH):
+        full_path = os.path.join(ADDON_DATA_IMG_PATH, file)
+        if os.path.isfile(full_path):
+            stat = os.stat(full_path)
+            if stat.st_mtime <= time_in_secs:
+                os.remove(full_path)
+
+    ''' Deletes old temporary files on startup
+    '''
+    for file in os.listdir(ADDON_DATA_IMG_TEMP_PATH):
+        full_path = os.path.join(ADDON_DATA_IMG_TEMP_PATH, file)
+        if os.path.isfile(full_path):
+            os.remove(full_path)
+
+
+def addon_data(file,content=None):
+    targetfile = os.path.join(ADDON_DATA_PATH, file)
+
+    if not content:
+        if not xbmcvfs.exists(targetfile):
+            return ''
+
+        try:
+            f = open(targetfile,'r')
```

Comment:
Use the python with statement to ensure the fd is always closed

Reply:
Sorry, don't get what you mean. It is closed after lines were read?

Reply:
In case of an exception in the try block the file descriptor is not closed. So normally we call f.close() in a finally: block or even better, we use the with statement:

```with open(x, 'r') as f:
         f.do_something()
```

Reply:
Perhaps also use the io.open() be more Python3 compatible.

Reply:
open() in Python3 is using io.open() afaik

Reply:
That's what I have found a while ago:

> In Python 2, the built-in open and io.open were different (io.open was newer and supported more things). In Python 3, open and io.open are now the same thing (they got rid of the old built-in open), so you should always use open. Code that needs to be compatible with Python 2 and 3 might have a reason to use io.open.

Reply:
@enen92  changed to descriptor usage. thanks for the hint.

Reply:
So your quote says
>Code that needs to be compatible with Python 2 and 3 might have a reason to use `io.open`
As we are looking for PY2/PY3 compatible add-ons, I would say that for a next release you look into `io.open`